File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        directions = [[1,0], [-1,0], [0,1], [0,-1]]
        ROWS = len(grid)
        COLS = len(grid[0])
        maxArea = 0

        def dfs(r, c):
            if (r < 0 or c < 0 or r >= ROWS or c >= COLS or grid[r][c] == 0):
                return 0

            grid[r][c] = 0  # mark visited
            area = 1

            for dr, dc in directions:
                area += dfs(r + dr, c + dc)

            return area

        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == 1:
                    maxArea = max(maxArea, dfs(row, col))

        return maxArea

File: Data Structures & Algorithms/count-number-of-islands/submission-0.swift
class Solution {
    func numIslands(_ grid: [[Character]]) -> Int {
        let directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        let ROWS = grid.count
        let COLS = grid[0].count
        var islands = 0
        var grid = grid

        func dfs(_ r: Int, _ c: Int) {
            if r < 0 || c < 0 || r >= ROWS || c >= COLS || grid[r][c] == "0" {
                return
            }

            grid[r][c] = "0"

            for dir in directions {
                dfs(r + dir[0], c + dir[1])
            }
        }

        for r in 0..<ROWS {
            for c in 0..<COLS {
                if grid[r][c] == "1" {
                    dfs(r, c)
                    islands += 1
                }
            }
        }
        
        return islands
    }
}
